eth/handler.go

```diff
@@ -191,12 +198,24 @@ func (pm *ProtocolManager) Stop() {
 	// take between 5 to 10 seconds and therefor feedback is required.
 	glog.V(logger.Info).Infoln("Stopping ethereum protocol handler...")
 
-	pm.quit = true
 	pm.txSub.Unsubscribe()         // quits txBroadcastLoop
 	pm.minedBlockSub.Unsubscribe() // quits blockBroadcastLoop
-	close(pm.quitSync)             // quits syncer, fetcher, txsyncLoop
 
-	// Wait for any process action
+	// Quit the sync loop.
+	// After this send has completed, no new peers will be accepted.
+	pm.noMorePeers <- struct{}{}
+
+	// Quit fetcher, txsyncLoop.
+	close(pm.quitSync)
+
+	// Disconnect existing sessions.
+	pm.peers.lock.Lock()
+	for _, peer := range pm.peers.peers {
+		peer.Disconnect(p2p.DiscQuitting)
+	}
+	pm.peers.lock.Unlock()
```

Comment:
I agree with @zsfelfoldi here. You might empty the peer set here but have in flight peers pending to be added, which would lead to deadlocks. I think a better solution is to add a Stop method to the `peerSet`, which drops all peers per the above code, but also sets a closed status on the set. If the closed status is set, the `peerSet.Register` would return an error.